src/map/clif.cpp

```diff
@@ -5754,30 +5754,39 @@ void clif_skillup(map_session_data *sd, uint16 skill_id, int lv, int range, int
 }
 
 
-/// Updates a skill in the skill tree (ZC_SKILLINFO_UPDATE2).
-/// 07e1 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B
-void clif_skillinfo(map_session_data *sd,int skill_id, int inf)
-{
-	nullpo_retv(sd);
-
-	const int fd = sd->fd;
+/// Updates a skill in the skill tree
+/// 07e1 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B (ZC_SKILLINFO_UPDATE2)
+/// 0b33 <skill id>.W <type>.L <level>.W <sp cost>.W <attack range>.W <upgradable>.B <level2>.W (ZC_SKILLINFO_UPDATE3)
+void clif_skillinfo(map_session_data& sd,int skill_id, int inf){
+#if PACKETVER >= 20090715
 	uint16 idx = skill_get_index(skill_id);
 
-	if (!session_isActive(fd) || !idx)
+	if (!session_isActive(sd.fd) || !idx)
 		return;
 
-	WFIFOHEAD(fd,packet_len(0x7e1));
-	WFIFOW(fd,0) = 0x7e1;
-	WFIFOW(fd,2) = skill_id;
-	WFIFOL(fd,4) = inf?inf:skill_get_inf(skill_id);
-	WFIFOW(fd,8) = sd->status.skill[idx].lv;
-	WFIFOW(fd,10) = skill_get_sp(skill_id,sd->status.skill[idx].lv);
-	WFIFOW(fd,12) = skill_get_range2(&sd->bl,skill_id,sd->status.skill[idx].lv,false);
-	if( sd->status.skill[idx].flag == SKILL_FLAG_PERMANENT )
-		WFIFOB(fd,14) = (sd->status.skill[idx].lv < skill_tree_get_max(skill_id, sd->status.class_))? 1:0;
+	PACKET_ZC_SKILLINFO_UPDATE2 p{};
+
+	p.packetType = HEADER_ZC_SKILLINFO_UPDATE2;	
+	p.id = skill_id;
+	p.level = sd.status.skill[idx].lv;
+	p.sp = skill_get_sp(skill_id,sd.status.skill[idx].lv);
+	p.range2 = skill_get_range2(&sd.bl,skill_id,sd.status.skill[idx].lv,false);
+	p.upFlag = 0;
+
+	if(inf >= 0)
+		p.inf = inf;
```

Comment:
upFlag is defined after
'inf' is always 0 in our code
```suggestion
	p.sp = static_cast<uint16>( skill_get_sp(skill_id,sd.status.skill[idx].lv) );
	p.range2 = static_cast<uint16>( skill_get_range2(&sd.bl,skill_id,sd.status.skill[idx].lv,false) );
```